src/bridge/cockpitpcpmetrics.c

```diff
@@ -561,60 +684,24 @@ convert_metric_description (CockpitPcpMetrics *self,
   return TRUE;
 }
 
-static void
-cockpit_pcp_metrics_prepare (CockpitChannel *channel)
+static gboolean
+prepare_current_context (CockpitPcpMetrics *self)
 {
-  CockpitPcpMetrics *self = COCKPIT_PCP_METRICS (channel);
-  const gchar *problem = "protocol-error";
+  gboolean result = FALSE;
   JsonObject *options;
-  const gchar *source;
   gchar **instances = NULL;
   gchar **omit_instances = NULL;
   JsonArray *metrics;
-  const char *name;
-  int type;
   int i;
 
-  COCKPIT_CHANNEL_CLASS (cockpit_pcp_metrics_parent_class)->prepare (channel);
+  g_free (self->metrics);
+  g_free (self->pmidlist);
 
-  options = cockpit_channel_get_options (channel);
+  self->numpmid = 0;
+  self->metrics = NULL;
+  self->pmidlist = NULL;
 
-  /* "source" option */
-  if (!cockpit_json_get_string (options, "source", NULL, &source))
-    {
-      g_warning ("invalid \"source\" option for metrics channel");
-      goto out;
-    }
-  else if (!source)
-    {
-      g_warning ("no \"source\" option specified for metrics channel");
-      goto out;
-    }
-  else if (g_str_equal (source, "direct"))
-    {
-      type = PM_CONTEXT_LOCAL;
-      name = NULL;
-    }
-  else if (g_str_equal (source, "pmcd"))
```

Comment:
> See [man pcpintro], $PMCD_CONNECT_TIMEOUT.

This process is async event driven cannot block for more than about max 100 ms anywhere during normal operation, and less than that is recommended.. If we reduce the timeout to less than 100 ms, then the connection to pmcd will likely race and fail. So we would need to do pmcd access out of process and/or in a thread.

So long running tasks like pmNewContext(PM_CONTEXT_HOST) will need